implementation.md

```diff
@@ -0,0 +1,336 @@
+1. Write a class that implements `DumperSource`. If a dumper by the requested
+name is available, return an instance wrapped by `Optional.of`, otherwise,
+`Optional.empty()`.
+1. Annotate this class with `@MetaInfServices`.
+1. Include a no-arguments constructor for the `DumperSource` class.
+
+Dumpers will get an array of values, one for each of the expressions provided
+by the user, boxed as `Object`. They must unbox the objects appropriately for
+themselves.
+
+### Input Format
+Creating a new source format is meant to be easy, but convoluted in order to be
+type safe. To create a new source format:
+
+1. Create a class, _V_, that will hold all the data for a “row” in the incoming
+data. It must be a class and not an interface.
+1. Create a parameterless method in _V_ for every variable to be exposed. The
+method names must be valid Shemsu names (lowercase with underscores) and
+decorated with `@Export` annotations with the correct type signature. All
+methods must return `boolean`, `long`, `String`, `Instant`, `Set`, or `Tuple`
+(and `Set` and `Tuple` may only contain more of the same).
+1. Create a new interface, _R_, extending `InputRepository<`_V_`>`.
+1. Create a new class that extends `BaseInputFormatDefinition<`_V_`,`_R_`>` and
+provides those types to the constructor as well as a name that will be used in
+the `Input` instruction. This class must be annotated with
+`@MetaInfServices(InputFormatDefinition)`.
+1. Create new classes that provide data which implement _R_ and are annotated
+with `@MetaInfServices(`_R_`)`.
+
+### Constants
+A constant is a value that can be generate with no input. It need not actually
+be constant (_e.g._, `now` is a constant). The Shesmu compiler will arbitrarily
+copy the value of a constant, so a constant should be side-effect free.
+
+1. Create a class that implements `ConstantSource`.
+1. Annotate this class with `@MetaInfServices`.
+1. Include a no-arguments constructor for the `ConstantSource` class.
+1. Return a stream of `Constant` objects for the constants to be created.
+
+There are a number of convience methods to create constants for values. If
+something more complicated is desired, extend the `Constant` class and write
+arbitrary bytecode in the `load` method. The code must not change any value on
+the stack and must leave exactly one value of the correct type on the stack.
+
+### Functions
+A function is a transformation of input data. It matches with a call to a
+method. Since there is no user-defined error-handling, these functions should
+not throw. Also, since Shesmu has no null values, they should not return null
+on an error.
+
+To provide functions:
+
+1. Create a class taht implements `FunctionRepository`.
+1. Annotate this class with `@MetaInfServices`.
+1. Include a no-arguments constructor for the `FunctionRepository` class.
+1. Return a stream of `FunctionDefintion` objects for the functions to be created.
+
+There are a few methods to generate a `FunctionDefinition`:
```

Comment:
`ways` instead of `methods`? You do use `method` technically elsewhere in the document.